server/games/game_rater.py

```diff
@@ -0,0 +1,101 @@
+import trueskill
+from trueskill import Rating
+from server.rating import RatingType
+from ..decorators import with_logger
+
+from server.games.game_results import GameOutcome
+from server.config import FFA_TEAM
+
+from typing import List, Dict, Iterable
+from server.players import  Player
+
+
+class GameRatingError(Exception):
+    pass
+
+
+@with_logger
+class GameRater(object):
+    def __init__(self,
+                players_by_team: Dict[int, List[Player]],
+                outcome_by_player: Dict[Player, GameOutcome],
+                rating_type=RatingType.GLOBAL):
+        self._rating_type = rating_type
+        self._outcome_by_player = outcome_by_player
+        self._players_by_team = players_by_team
+
+
+    def compute_rating(self) -> Dict[Player, Rating]:
+        rating_groups = self._get_rating_groups()
+        team_outcomes = [self._get_team_outcome(team.keys()) for team in rating_groups]
+        ranks = self._ranks_from_team_outcomes(team_outcomes)
+
+        self._logger.debug("Rating groups: %s", rating_groups)
+        self._logger.debug("Ranks: %s", ranks)
+        return trueskill.rate(rating_groups, ranks)
+
+
+    def _get_rating_groups(self) -> List[Dict[Player, Rating]]:
+        """
+        Converts a dictionary mapping team ids to players to the trueskill rating_group format
+        example input: {team1: [p1, p2], team2: [p3, p4]}
+        example output: [ {p1: Rating, p2: Rating}, {p3: Rating, p4: Rating} ]
+        """
+        if FFA_TEAM in self._players_by_team:
+            if len(self._players_by_team[FFA_TEAM]) is 2 and len(self._players_by_team) is 1:
+                return [
+                    {player: Rating(*player.ratings[self._rating_type])}
+                    for player in self._players_by_team[FFA_TEAM]
+                ]
+            else:
+                raise GameRatingError(
+                    f"Attempted to rate FFA game with other than two players: {{team: players}} = {self._players_by_team}"
+                )
+        elif len(self._players_by_team) is 2:
```

Comment:
Always use `==` for numerical comparison. Sometimes `is` will work as you expect, but it is by no means guaranteed and can even vary in different python implementations. An example in cpython on my laptop:

```python
>>> 20**10 is 20**10
False
>>> 20**10 == 20**10
True
```

`is` is for object identity when you want to make sure that you actually have a reference to the same object in memory, and not just a similar object which has the same value:

```python
>>> [] is []
False
>>> a = []
>>> b = a
>>> a is b
True
```

Booleans should always be compared with `is` because in python `True`, `False` and `None` are all singleton objects. Writing `a = True` simply makes `a` a reference to `True`.